e2e/setup.ts

```diff
@@ -34,8 +48,10 @@ export default async (): Promise<void> => {
 	server.listen(port, host);
 	globalThis.server = server;
 
-	const dom = new JSDOM();
-	globalThis.document = dom.window.document;
+	const browser = await launch({ dumpio: true });
+	globalThis.__BROWSER_GLOBAL__ = browser;
```

Comment:
https://github.com/akashic-games/pdi-browser/pull/252#discussion_r1006580734 のコメントにもある通り公式ドキュメントを参考にしただけです。

Reply:
jest の test-environment については [こちら](https://jestjs.io/ja/docs/configuration#testenvironment-string) が参考になるかと思います。

質問の断片的な答えとしては以下に該当するかと。

> You can also pass variables from this module to your test suites by assigning them to this.global object – this will make them available in your test suites as global variables.

Reply:
あとは [こちら](https://jestjs.io/ja/docs/configuration#globalsetup-string) でしょうか。

> Any global variables that are defined through globalSetup can only be read in globalTeardown. You cannot retrieve globals defined here in your test suites.

Reply:
不勉強で大変申し訳ないのですが、あまり理解できませんでした。。
直感的には`puppeteer.launch`と`puppeteer.connect`を両方利用せず片方だけでいいのではと思っているのですが、それだと問題がある感じなのでしょうか？

Reply:
できればもう少しご自分で調べていただきたいのですが、以下が私の認識です。

> すみませんが、this.global.__BROWSER_GLOBAL__と分けて定義している理由とそれぞれ異なるものが代入されている理由を教えていただきたいです

* `globalSetup`, `globalTeardown` で定義された `globalThis` は各テストからは参照できない [1](https://jestjs.io/ja/docs/configuration#globalsetup-string)
* jest-environment の `this.global` に代入された値は各テストの `globalThis` で参照することができる [2](https://jestjs.io/ja/docs/configuration#testenvironment-string)

したがってそれを解決した結果が理由です。

> 直感的にはpuppeteer.launchとpuppeteer.connectを両方利用せず片方だけでいいのではと思っているのですが、それだと問題がある感じなのでしょうか？

[PuppeteerNode.launch()](https://pptr.dev/api/puppeteer.puppeteernode.launch/) はブラウザインスタンスの起動、[Puppeteer.connect()](https://pptr.dev/api/puppeteer.puppeteer.connect/) はブラウザインスタンスにアタッチ (ここでは WebSocket 経由) するものなので、全くの別物です。